src/types/query.d.ts

```diff
@@ -1,28 +1,109 @@
+import Expr, {
+  ExprVal,
+  Lambda,
+  Document,
+  Collection,
+  Page,
+  Index,
+  Function,
+} from './Expr'
 
-type ExprVal = Expr | string | number | boolean | { [key: string]: any }
-type ExprArg = ExprVal | Array<ExprVal>
-export type Lambda = (...vars: any[]) => Expr
+export type ExprArg<T = unknown> = ExprVal<T> | Array<ExprVal<T>>
+
+export type CreateParams<T> = ExprVal<{
+  data: T
+  credentials?: object
+  delegates?: object
+  ttl?: Expr.Time
+}>
+
+export type CreateCollectionParams = {
+  name: string
+  data?: object
+  permissions?: object
+  history_days?: number
+  ttl_days?: number
+}
+
+export type CreateIndexParams = {
+  name: string
+  source: Expr.CollectionRef<any> | any[]
+  terms?: any[]
+  values?: any[]
+  unique?: boolean
+  serialized?: boolean
+  permissions?: object
+  data?: object
+}
+
+export type PaginateParams = {
+  ts?: number | Expr.Time
+  size?: number
+  before?: string
+  after?: string
+  events?: boolean
+  sources?: boolean
+}
+
+export type FunctionParams = {
+  name: string
+  body: object
+  data?: object
+  role?: any
+}
 
 export module query {
-  export function Ref(ref: ExprArg, id?: ExprArg): Expr
+  export function Ref<T>(
+    ref: Expr.CollectionRef<T>,
+    id?: ExprVal<number | string>
+  ): Expr.DocumentRef<T>
+
+  // TODO
   export function Bytes(bytes: ExprArg | ArrayBuffer | Uint8Array): Expr
   export function Abort(msg: ExprArg): Expr
   export function At(timestamp: ExprArg, expr: ExprArg): Expr
-  export function Let(vars: ExprArg, in_expr: ExprArg): Expr
+
+  export function Let<T>(vars: ExprVal<object>, in_expr: Expr<T>): Expr<T>
+
+  // TODO
   export function Var(varName: ExprArg): Expr
-  export function If(
-    condition: ExprArg,
-    then: ExprArg | null,
-    _else: ExprArg | null
-  ): Expr
+
+  export function If<T>(
+    condition: ExprArg<boolean>,
+    then: ExprArg<T>,
+    _else: ExprArg<T>
+  ): Expr<T>
+
+  export function If<T, U>(
+    condition: ExprArg<boolean>,
+    then: ExprArg<T>,
+    _else: ExprArg<U>
+  ): Expr<T | U>
+
+  // TODO
   export function Do(...args: ExprArg[]): Expr
   export function Object(fields: ExprArg): Expr
-  export function Lambda(f: Lambda): Expr
-  export function Lambda(var_name: ExprArg, expr: ExprArg): Expr
-  export function Call(ref: ExprArg, ...args: ExprArg[]): Expr
+
+  export function Lambda<In extends any[], Out>(
+    f: (...args: In) => Out
+  ): Expr<Lambda<In, Out>>
+
+  export function Lambda<In extends any[], Out>(
+    var_name: ExprArg<string>,
+    expr: ExprArg<Out>
+  ): Expr<Lambda<In, Out>>
+
+  export function Call<T>(ref: Expr.FunctionRef<T>, ...args: any[]): Expr<T>
+
+  // TODO
   export function Query(lambda: ExprArg | Lambda): Expr
-  export function Map(collection: ExprArg, lambda_expr: ExprArg | Lambda): Expr
+
+  export function Map<T, Out>(
+    collection: Expr.Iterable<T>,
+    lambda_expr: ExprVal<Lambda<[T extends Expr ? T : Expr<T>], Out>>
+  ): Expr<Out[]>
```

Comment:
the output wound't be `Expr<Iterable<Out>>`?

Reply:
depends on what you expect `client.query(q.Map(...))` to resolve to.

if the mapping function returns `Expr<number>`, I would expect `Promise<number[]>` to be returned by `client.query`

Reply:
`q.Map` would never return anything different than an Iterable

Reply:
Not sure what you mean. When doesn't `q.Map` return an array?

```js
const result = await db.query(q.Map([1], x => q.Add(x, 1)))
result // => [ 2 ]
```

Reply:
I said it always will return an iterable, if you pass an array it will return an array, if you pass a page it will return a page. Map doesnt support sets yet but should in the future

Reply:
Pages
```javascript
> Paginate(Databases(), {size: 4})
{
  after: [ Database("blah") ],
  data: [
    Database("blah_renamed"),
    Database("db1"),
    Database("db2"),
    Database("test_driver")
  ]
}
```

Map+Pages
```javascript
> Map(Paginate(Databases(), {size: 4}), Lambda(ref => Get(ref)))
{
  after: [ Database("blah") ],
  data: [
    {
      ref: Database("blah_renamed"),
      ts: 1578425890150000,
      priority: 1,
      name: 'blah_renamed'
    },
    {
      ref: Database("db1"),
      ts: 1583508872680000,
      name: 'db1',
      data: { gql: { ts: Time("2020-03-06T15:34:32.296894Z") } }
    },
    { ref: Database("db2"), ts: 1563914425250000, name: 'db2' },
    {
      ref: Database("test_driver"),
      ts: 1569504687220000,
      name: 'test_driver'
    }
  ]
}
```

Reply:
Fixed 👍